Refactor arguments + environment variables

soyuka opened the pull request:
1. Implement `minimist` for cli options (`--host`, `--port`, `--threads` etc.)
2. Review `puppeteer` constructor arguments and `window.init`, use an object instead of argument list
3. Implement `--threads` 

fix #7 

My next step would be to implement some pretty logging that looks just like the miner on https://coin-hive.com/ homepage. 

ping @tailot  WDYT?

cazala commented:
Looking great! The only thing I would change is that I would set the default `THREADS` to something like `null` or `-1` or whatever value that makes more sense for you, and when that value is passed to `window.init` I would do something like:

```js
if (threads != null) {
  miner.setNumThreads(threads);
}
```

I would do it this way because by default the miner uses `navigator.hardwareConcurrency` to find the number of CPU cores.

tailot commented:
@soyuka Good job! 
You could join pieces like these:
``` javascript
  const interval = argv.interval || defaults.INTERVAL || process.env.COINHIVE_INTERVAL;
  const port = argv.port || defaults.PORT || process.env.COINHIVE_PORT;
  const host = argv.host || defaults.HOST || process.env.COINHIVE_HOST;
  const threads = argv.threads || defaults.THREADS || process.env.COINHIVE_THREADS;

  const miner = await CoinHive(siteKey, {interval, port, host, threads});
```
in
``` javascript
  const options = {
    interval: argv.interval || defaults.INTERVAL || process.env.COINHIVE_INTERVAL,
    port: argv.port || defaults.PORT || process.env.COINHIVE_PORT,
    host: argv.host || defaults.HOST || process.env.COINHIVE_HOST,
    threads: argv.threads || defaults.THREADS || process.env.COINHIVE_THREADS
  }
  const miner = await CoinHive(siteKey, options);
```
If you add a parameter, make a single change.

soyuka commented:
Thanks for the review guys, updated.

Interestingly, using `setThread` was starting the miner. I've changed tests back to what they were but maybe it's worth mentioning? 

I've also updated the README  with detailed instructions.

cazala commented on src/index.js:
```diff
@@ -2,31 +2,24 @@ const server = require('./server');
-  port = defaults.PORT,
-  host = defaults.HOST
-) {
+module.exports = async function getRunner(siteKey, options = {port: defaults.PORT, host: defaults.HOST, interval: defaults.INTERVAL, threads: defaults.THREADS}) {
```
This will fail when the user sends a partial options object. ie:

```js
function withDefaults(options = { host: 'localhost', port: 8080}){ 
  console.log(options.host + ':' + options.port);
}
withDefaults() // localhost:8080
withDefaults({ port: 8081 }) // undefined:8081
```

I think we should do something like this:

```js
const defaultOptions = {
  port: defaults.PORT, 
  host: defaults.HOST, 
  interval: defaults.INTERVAL, 
  threads: defaults.THREADS
};
module.exports = async function getRunner(siteKey, constructorOptions = defaultOptions) {
  const options = Object.assign({}, defaultOptions, constructorOptions);
  ...
```

soyuka replied:
:+1: changed options keys
